pkg/utils/grpc.go

```diff
@@ -27,44 +46,139 @@ func NewGrpc(serverType string, port int, certPem, keyPem, rootCertPem []byte) (
 
 	log.Debug().Msgf("Parameters for %s gRPC server: MaxConcurrentStreams=%d;  KeepAlive=%+v", serverType, maxStreams, streamKeepAliveDuration)
 
+	s := &Server{
+		name:     serverType,
+		cm:       cm,
+		certName: certName,
+	}
+
 	grpcOptions := []grpc.ServerOption{
 		grpc.MaxConcurrentStreams(maxStreams),
 		grpc.KeepaliveParams(keepalive.ServerParameters{
 			Time: streamKeepAliveDuration,
 		}),
 	}
 
-	mutualTLS, err := setupMutualTLS(false, serverType, certPem, keyPem, rootCertPem)
-	if err != nil {
-		log.Error().Err(err).Msg("Error setting up mutual tls for GRPC server")
-		return nil, nil, err
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         serverType,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		MinVersion:         constants.MinTLSVersion,
+		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			return &s.config, nil
+		},
 	}
+	mutualTLS := grpc.Creds(credentials.NewTLS(&tlsConfig))
 	grpcOptions = append(grpcOptions, mutualTLS)
 
-	return grpc.NewServer(grpcOptions...), lis, nil
+	s.server = grpc.NewServer(grpcOptions...)
+	return s, lis, nil
+}
+
+// GetServer returns the gRPC server
+func (s *Server) GetServer() *grpc.Server {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.server
+}
+
+func (s *Server) setConfig() error {
```

Comment:
A setter method without any arguments seems ambiguous. Would rename this to `initConfig` or something similar.